src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -0,0 +1,1036 @@
+                        _inBytesPacket = ((int)_partialHeaderBuffer[TdsEnums.HEADER_LEN_FIELD_OFFSET] << 8 |
+                                  (int)_partialHeaderBuffer[TdsEnums.HEADER_LEN_FIELD_OFFSET + 1]) - _inputHeaderLen;
+
+                        _messageStatus = _partialHeaderBuffer[1];
+                        _spid = _partialHeaderBuffer[TdsEnums.SPID_OFFSET] << 8 |
+                                  _partialHeaderBuffer[TdsEnums.SPID_OFFSET + 1];
+#if !NETFRAMEWORK
+                        SqlClientEventSource.Log.TryAdvancedTraceEvent("TdsParserStateObject.TryProcessHeader | ADV | State Object Id {0}, Client Connection Id {1}, Server process Id (SPID) {2}", _objectID, _parser?.Connection?.ClientConnectionId, _spid);
+#endif
+                    }
+                    else
+                    {
+                        Debug.Assert(_inBytesUsed == _inBytesRead, "Did not use all data while reading partial header");
+
+                        // Require more data
+                        if (_parser.State == TdsParserState.Broken || _parser.State == TdsParserState.Closed)
+                        {
+                            // NOTE: ReadNetworkPacket does nothing if the parser state is closed or broken
+                            // to avoid infinite loop, we raise an exception
+                            ThrowExceptionAndWarning();
+                            return true;
+                        }
+
+                        if (!TryReadNetworkPacket())
+                        {
+                            return false;
+                        }
+
+                        if (IsTimeoutStateExpired)
+                        {
+                            ThrowExceptionAndWarning();
+                            return true;
+                        }
+                    }
+                } while (_partialHeaderBytesRead != 0); // This is reset to 0 once we have read everything that we need
+
+                AssertValidState();
+            }
+            else
+            {
+                // normal header processing...
+                _messageStatus = _inBuff[_inBytesUsed + 1];
+                _inBytesPacket = (_inBuff[_inBytesUsed + TdsEnums.HEADER_LEN_FIELD_OFFSET] << 8 |
+                                              _inBuff[_inBytesUsed + TdsEnums.HEADER_LEN_FIELD_OFFSET + 1]) - _inputHeaderLen;
+                _spid = _inBuff[_inBytesUsed + TdsEnums.SPID_OFFSET] << 8 |
+                                              _inBuff[_inBytesUsed + TdsEnums.SPID_OFFSET + 1];
+#if !NETFRAMEWORK
+                SqlClientEventSource.Log.TryAdvancedTraceEvent("TdsParserStateObject.TryProcessHeader | ADV | State Object Id {0}, Client Connection Id {1}, Server process Id (SPID) {2}", _objectID, _parser?.Connection?.ClientConnectionId, _spid);
+#endif
+                _inBytesUsed += _inputHeaderLen;
+
+                AssertValidState();
+            }
+
+            if (_inBytesPacket < 0)
+            {
+#if NETFRAMEWORK
```

Comment:
If the netfx version provides better information than the netcore version we should use the netfx approach. Can you trace down whether these cause different messages to be output and if they do identify which is more useful in terms of reporting the error?